### PR TITLE
fix(plugin-docs,theme): refactor docs plugin routes and component tree

### PR DESCRIPTION
**Pull request by slorber**
## Breaking changes

This refactor significantly changes the docs theme component tree. If you swizzled `@theme/DocPage`, or doc tags pages, you will likely need to upgrade your theme code a bit.

The tree of theme components rendering doc routes changed from:

```bash
Root
├── @theme/DocPage
│   └── @theme/DocPage/Layout
│       ├── @theme/DocPage/Layout/Sidebar
│       └── @theme/DocPage/Layout/Main
│           ├── @theme/DocCategoryGeneratedIndexPage
│           └── @theme/DocItem
├── @theme/DocTagsListPage
└── @theme/DocTagDocListPage
```

to this new structure

```bash
Root
└── @theme/DocsRoot
    └── @theme/DocVersionRoot
        ├── @theme/DocRoot
        │   └── @theme/DocRoot/Layout
        │       ├── @theme/DocRoot/Layout/Sidebar
        │       └── @theme/DocRoot/Layout/Main
        │           ├── @theme/DocCategoryGeneratedIndexPage
        │           └── @theme/DocItem
        ├── @theme/DocTagsListPage
        └── @theme/DocTagDocListPage
```

- `@theme/DocPage` has been renamed to `@theme/DocRoot`
- `@theme/DocRoot` (renamed) does not receive a `versionMetadata` prop anymore. The new `@theme/DocVersionRoot` receives it (upper in the tree), and makes it available in its subtree with the `useDocsVersion()` hook.
- `@theme/DocRoot` (renamed),  `@theme/DocTagsListPage` and `@theme/DocTagDocListPage` do not apply `<Layout>` anymore: it's applied in `@theme/DocsLayout` (parent of all docs plugin routes)
- option `docPageLayoutComponent` has been renamed to `docRootComponent`

Note there's a clear distinction between:
- `@theme/DocsRoot`: top-level parent layout component that wraps every docs plugin routes (all versions, docs, generated index, docs tags pages...) and apply the `<Layout>` globally
- `@theme/DocRoot`: single doc layout, responsible for displaying the docs sidebar alongside the docs content (markdown or category generated index)

## Motivation

All the versioned docs pages should have a shared parent layout page. This prevents unmounting/remounting the layout when navigating across docs pages, category-generated index pages, tags pages, and even across versions.

New docs plugin options:
- `docsRootComponent`: the component that wraps every route of a single doc plugin instance
- `docVersionRootComponent`: the component that wraps every route of a specific doc version

With this refactor, tags-related pages are able to use the `useDocsVersion()` hook.

Also fixes a bug in the core route sorting algorithm which was not really recursive.

## Test Plan

unit tests + preview

### Test links

Prod and deploy preview should look exactly the same.

Doc:
- https://deploy-preview-7966--docusaurus-2.netlify.app/docs/installation
- https://docusaurus.io/docs/installation

Tags list:
- https://deploy-preview-7966--docusaurus-2.netlify.app/tests/docs/tags
- https://docusaurus.io/tests/docs/tags

Tag page:
- https://deploy-preview-7966--docusaurus-2.netlify.app/tests/docs/tags/some-tag
- https://docusaurus.io/tests/docs/tags/some-tag

### CODE REVIEW

**Comment by Josh-Cena** [comment]
Does this supercede #6823?

File: packages/docusaurus-plugin-content-docs/src/__tests__/options.test.ts [hunk 1]
```diff
@@ -49,7 +49,9 @@ describe('normalizeDocsPluginOptions', () => {
       sidebarPath: 'my-sidebar', // Path to sidebar configuration for showing a list of markdown pages.
       sidebarItemsGenerator: DefaultSidebarItemsGenerator,
       numberPrefixParser: DefaultNumberPrefixParser,
-      docLayoutComponent: '@theme/DocPage',
+      docsRootComponent: '@theme/DocsRoot',
+      docVersionRootComponent: '@theme/DocVersionRoot',
+      docRootComponent: '@theme/DocRoot',
```
Comment (slorber) [inline packages/docusaurus-plugin-content-docs/src/__tests__/options.test.ts:54]: New Root theme component we decided with @Josh-Cena .

`@theme/DocPage` becomes `@theme/DocRoot`

Hierarchy is: `@theme/DocsRoot` > `@theme/DocVersionRoot` > `@theme/DocRoot`

**Review by slorber — COMMENTED** [review]
Ready to review 👍

**Review by Josh-Cena — CHANGES_REQUESTED** [review]
On https://deploy-preview-7966--docusaurus-2.netlify.app/docs/tags/:

image

`NotFound` renders `Layout` already

**Comment by slorber** [comment]
Thanks, good catch 😅 wonder how I'll fix that

**Comment by Josh-Cena** [comment]
Can it be a prop in `NotFound` to not render `Layout`?

**Pull request merged**